nucliadb/src/nucliadb/ingest/orm/knowledgebox.py

```diff
@@ -80,78 +85,140 @@ def new_unique_kbid() -> str:
+                            semantic_model.matryoshka_dimensions  # type: ignore
+                        )
+                    kb_shards.similarity = semantic_model.similarity_function
+                    kb_shards.model.CopyFrom(semantic_model)
+
+                else:
+                    for vectorset_id, semantic_model in semantic_models.items():  # type: ignore
+                        # if this KB uses a matryoshka model, we can choose a different
+                        # dimension
+                        if len(semantic_model.matryoshka_dimensions) > 0:
+                            dimension = choose_matryoshka_dimension(semantic_model.matryoshka_dimensions)
+                        else:
+                            dimension = semantic_model.vector_dimension
+
+                        vectorset_config = knowledgebox_pb2.VectorSetConfig(
+                            vectorset_id=vectorset_id,
+                            vectorset_index_config=nodewriter_pb2.VectorIndexConfig(
+                                similarity=semantic_model.similarity_function,
+                                # XXX: hardcoded value
+                                vector_type=nodewriter_pb2.VectorType.DENSE_F32,
+                                normalize_vectors=len(semantic_model.matryoshka_dimensions) > 0,
+                                vector_dimension=dimension,
+                            ),
+                            matryoshka_dimensions=semantic_model.matryoshka_dimensions,
+                        )
+                        await datamanagers.vectorsets.set(txn, kbid=kbid, config=vectorset_config)
+
+                await datamanagers.cluster.update_kb_shards(txn, kbid=kbid, shards=kb_shards)
+
+                # shard creation will alter this value on maindb, make sure nobody
+                # uses this variable anymore
+                del kb_shards
+
+                # Create in storage
+
+                storage = await get_storage(service_name=SERVICE_NAME)
+
+                created = await storage.create_kb(kbid)
+                rollback_ops.append(partial(storage.delete_kb, kbid))
+                if not created:
+                    logger.error(f"KB {kbid} could not be created")
+                    raise KnowledgeBoxCreationError(
+                        f"KB blob storage could not be created (slug={slug})"
+                    )
+
+                # Create shards in index nodes
+
+                shard_manager = get_shard_manager()
+                # XXX creating a shard is a slow IO operation that requires a write
+                # txn to be open!
+                await shard_manager.create_shard_by_kbid(txn, kbid)
+                # shards don't need a rollback as they will be eventually purged
+
+                await txn.commit()
+
+        except Exception:
+            # rollback all changes on the db and raise the exception
+            for op in reversed(rollback_ops):
+                await op()
```

Comment:
wdyt about guarding each `await op()` inside a try/except with a logg.error + continue? so we attempt to rollback as much as possible but in the end raise the error we wanted to raise

Reply:
Sure, good idea!